teerex/src/lib.rs

```diff
@@ -717,11 +597,45 @@ impl<T: Config> Pallet<T> {
 			Err(<Error<T>>::RemoteAttestationTooOld.into())
 		}
 	}
-}
 
-impl<T: Config> OnTimestampSet<T::Moment> for Pallet<T> {
-	fn on_timestamp_set(moment: T::Moment) {
-		Self::unregister_silent_workers(moment)
+	pub fn poke_shard(
+		shard: ShardIdentifier,
+		enclave_signer: &T::AccountId,
+	) -> Result<ShardSignerStatuses<T>, DispatchErrorWithPostInfo> {
+		let enclave = Self::sovereign_enclaves(enclave_signer.clone())
+			.ok_or(<Error<T>>::EnclaveIsNotRegistered)?;
+
+		let current_block_number = <frame_system::Pallet<T>>::block_number();
+
+		let fresh_status = ShardSignerStatus {
+			signer: enclave_signer.clone(),
+			fingerprint: enclave.fingerprint(),
+			last_activity: current_block_number,
+		};
+
+		let signer_statuses = if let Some(mut status_vec) = <ShardStatus<T>>::get(shard) {
+			if let Some(index) = status_vec.iter().position(|i| i.signer == *enclave_signer) {
+				status_vec[index] = fresh_status;
+			} else {
+				status_vec.push(fresh_status)
+			}
+			status_vec
+		} else {
+			vec![fresh_status]
+		};
```

Comment:
```suggestion
		let signer_statuses = get_or_init(...);
		
// more efficient, and maybe even more readable. :)
pub fn get_or_init(...) {
  <ShardStatus<T>>::get(shard)
  .map(|status_vec| {
      status_vec.iter().position(|i| i.signer == *enclave_signer)
        .map_or_else(|| status_vec.push(free_status), |i| status_vec[i] = fresh_status);
      
      status_vec
  }).unwrap_or_else(|| vec![fresh_status])
}
```

Reply:
borrow checker not happy with this. but I adapted pragmatically ;-)